src/Symfony/Component/Security/Http/MagicLink/MagicLinkTokenStorageInterface.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\MagicLink;
+
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * A class that is able to store, find and invalidate tokens used in magic links.
+ *
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+interface MagicLinkTokenStorageInterface
+{
+    /**
+     * Stores all the token information using the selector as the lookup key.
+     *
+     * The key that is sent to to the user will normally be a combination
+     * of selector + the un-hashed verifier.
+     */
+    public function storeToken(string $selector, string $hashedVerifier, UserInterface $user, \DateTimeInterface $expiresAt): void;
+
+    /**
+     * Find and return the token information.
+     */
+    public function findToken(string $selector): ?StoredMagicLinkTokenInterface;
+
+    /**
+     * Invalidate the given token.
+     */
+    public function invalidateToken(string $selector): void;
+
+    /**
+     * Invalidate all tokens for the given user.
+     *
+     * This is called whenever a new token is generated to guarantee
+     * that only one valid token is ever active. This is an extra
+     * security mechanism.
```

Comment:
Is this really needed? I.e. I should not be able to login at my PC and mobile at the same time?

Reply:
No idea! But 2 things:

1) I stole this idea from Auth0's magic link implementation.
2) It's NOT that you can't login at your PC and mobile at the same time. It's just that you would need to generate 1 login link for your PC, click it (get logged in) and then generate another login link that you use for your mobile. The only problem would be (and I don't think this flow is realistic) if you generate 2 login links.. and THEN you try to use each one. You need to "generate", "use", "generate", "use"

Reply:
:+1: then. Let's follow our reference examples - they are the best guidelines at this moment